lib/deimos/active_record/batch_consumption.rb

```diff
@@ -0,0 +1,179 @@
+# frozen_string_literal: true
+
+module Deimos
+  module ActiveRecord
+    # Batch consume methods
+    module BatchConsumption
+      # Handle a batch of Kafka messages
+      # @param payloads [Array<Hash>] Decoded payloads.
+      # @param metadata [Hash] Information about batch, including keys.
+      def consume_batch(payloads, metadata)
+        # Create [key, value] pairs, even if no keys are decoded
+        messages = payloads.
+          zip(metadata[:keys]).
+          map { |p, k| [k, p] }
+
+        Deimos.instrument('ar_consumer.consume_batch', topic: metadata[:topic], messages: messages) do
+          slices = _slice_batch(messages)
+
+          # The entire batch should be treated as one transaction so that if
+          # any message fails, the whole thing is rolled back
+          _retry_deadlock(topic: metadata[:topic]) do
+            ::ActiveRecord::Base.transaction do
+              slices.each do |slice|
+                removed, upserted = slice.partition { |_, v| v.nil? }
+
+                upsert_records(upserted) if upserted.any?
+                remove_records(removed) if removed.any?
+              end
+            end
+          end
+        end
+      end
+
+      # Get unique key for the ActiveRecord instance from the incoming key.
+      # Override this method (with super) to customize the set of attributes that
+      # uniquely identifies each record in the database.
+      # @param key [String] The encoded key.
+      # @return [Hash] The key attributes.
+      def record_key(key)
+        decoded_key = decode_key(key)
+
+        if decoded_key.nil?
+          {}
+        elsif decoded_key.is_a?(Hash)
+          @key_converter.convert(decoded_key)
+        else
+          { @klass.primary_key => decoded_key }
+        end
+      end
+
+    protected
+
+      # Upsert any non-deleted records
+      # @param records [Array<Array>] List of [key, value] pairs for a group of
+      # non-tombstone records.
+      def upsert_records(records)
+        key_cols = key_columns(records)
+
+        messages = records.map do |k, v|
+          record_attributes(v, k).
+            merge(record_key(k))
+        end
```

Comment:
These are basically record hashes, not messages, right? These are things that will be imported.